core/src/main/java/club/mondaylunch/gatos/core/graph/Graph.java

```diff
@@ -296,14 +328,108 @@ public Optional<List<Node>> getExecutionOrder() {
      * @return whether the connection is valid for the given node
      */
     private static boolean isConnectionValid(Node node, NodeConnection<?> connection) {
-        if (connection.from().nodeId().equals(node.id())) {
-            return node.outputs().containsValue(connection.from());
+        if (connection.from().nodeId().equals(node.getId())) {
+            return node.getOutputs().containsValue(connection.from());
         }
 
-        if (connection.to().nodeId().equals(node.id())) {
-            return node.inputs().containsValue(connection.to());
+        if (connection.to().nodeId().equals(node.getId())) {
+            return node.getInputs().containsValue(connection.to());
         }
 
         return false;
     }
+
+    /**
+     * Gets the number of nodes in this graph.
+     *
+     * @return the number of nodes
+     */
+    public int nodeCount() {
+        return this.nodes.size();
+    }
+
+    /**
+     * Gets the number of connections in this graph.
+     *
+     * @return the number of connections
+     */
+    public int connectionCount() {
+        return this.connections.size();
+    }
+
+    /**
+     * Needed so that {@link #serializableNodes} will be serialized automatically.
+     *
+     * @return the serializable nodes.
+     */
+    @SuppressWarnings("unused")
+    @ApiStatus.Internal
+    public Set<Node> getSerializableNodes() {
+        return this.serializableNodes;
+    }
+
+    /**
+     * Needed so that {@link #serializableMetadata} will be serialized automatically.
+     *
+     * @return the serializable metadata.
+     */
+    @SuppressWarnings("unused")
+    @ApiStatus.Internal
+    public Map<String, NodeMetadata> getSerializableMetadata() {
+        return this.serializableMetadata;
+    }
+
+    @Override
+    public void beforeSerialization() {
```

Comment:
This is clever. I don't really like how much extra logic & state we're adding to Graph for the sake of serialisation though. I'm happy to write a custom codec.

Reply:
This exists because mongo is stupid, the provided map codec only supports maps with String keys

Reply:
I need to use this so that the default pojo codec works.